### PR TITLE
Improve Attachment API and Attachment Specs

### PR DESCRIPTION
**Pull request by jhmegorei**
This PR improves the Attachment API and adds a lot of specs for the attachment model and its included concerns.
Not everything is specced though, due to time constraints. I urge other developers to add to the specs, so that especially the AttachmentJcampAasm Concern gets completely tested (it uses the also untested Jcamp Interface)

### WARNING
This PR removes the following methods from the Attachment model:
- for_report?
- for_template?
- report_id
- report
- update_report!

### CODE REVIEW

File: app/api/chemotion/attachable_api.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# frozen_string_literal: true
+
+module Chemotion
+  class AttachableAPI < Grape::API
+    resource :attachable do
```
Comment (FabianMauz) [inline app/api/chemotion/attachable_api.rb:5]: Where is this resource used?

File: app/usecases/attachments/upload_chunk_complete.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+          filename: file_name,
+          key: params[:key],
+          file_path: file_path,
+          created_by: user.id,
+          created_for: user.id,
+          content_type: MIME::Types.type_for(file_name)[0].to_s
+        )
+        attachment.save!
+
+        true
```
Comment (FabianMauz) [inline app/usecases/attachments/upload_chunk_complete.rb:54]: Why always true, even if an attachment could not be saved?

**Review by JanCBrammer — APPROVED** [review]

File: spec/models/attachment_spec.rb [hunk 3]
```diff
@@ -3,102 +3,833 @@
...
+        expect(subject).to eq(true)
+      end
+    end
+  end
+
+  describe '#for_container?' do
+    subject { attachment.for_container? }
+
+    context 'when not attached to container' do
+      let(:attachment) { create(:attachment, :attached_to_research_plan) }
```
Comment (phuang26) [inline spec/models/attachment_spec.rb:110]: typo? attached_to_research_plan?

Comment (jhmegorei) [reply spec/models/attachment_spec.rb:110]: no this is exactly the point. The context describes an attachment NOT being attached to a container, so I created on attached to a research_plan (which is not a container)

File: app/usecases/attachments/unlink.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+      end
+
+      def self.execute!(attachment)
+        new(attachment).execute!
+      end
+
+      def execute!
+        attachment.tap do |a|
+          a.attachable_id = nil
+          a.attachable_type = 'Container'
```
Comment (phuang26) [inline app/usecases/attachments/unlink.rb:19]: why?

Comment (jhmegorei) [reply app/usecases/attachments/unlink.rb:19]: why what? can you be more specific please?

Comment (phuang26) [reply app/usecases/attachments/unlink.rb:19]: why the attachable_type set to 'Container' when unlink?

Comment (jhmegorei) [reply app/usecases/attachments/unlink.rb:19]: That is what we wondered as well... please see https://github.com/ComPlat/chemotion_ELN/commit/e04c83b047af65f528e965023341ba04b3a070cd

**Review by phuang26 — COMMENTED** [review]
the attachable_spec.rb were removed and they were introduced for "Report LSDF - Spec".

**Review by phuang26 — APPROVED** [review]

**Pull request merged**